src/webOAuthServer.ts

```diff
@@ -0,0 +1,374 @@
+      return authCode;
+    }
+    return null;
+  }
+
+  /**
+   * Closes the request
+   *
+   * @param request the http request
+   */
+  private closeRequest(request: WebOAuthServer.Request): void {
+    request.connection.end();
+    request.connection.destroy();
+  }
+
+  /**
+   * Validates that the state param in the auth url matches the state
+   * param in the http request
+   *
+   * @param request the http request
+   */
+  private validateState(request: WebOAuthServer.Request): boolean {
+    const state = request.query.state;
+    const query = parseUrl(this.authUrl, true).query;
+    return !!(state && state === query.state);
+  }
+}
+
+namespace WebOAuthServer {
+  export interface Options {
+    oauthConfig: OAuth2Options;
+  }
+
+  export type Request = http.IncomingMessage & { query: { code: string; state: string } };
+}
+
+/**
+ * Handles the actions specific to the http server
+ */
+class WebServer extends AsyncCreatable<WebServer.Options> {
+  private static DEFAULT_CLIENT_SOCKET_TIMEOUT = 20000;
+  public server!: http.Server;
+  public port = 1717;
+  public host = 'localhost';
+  private logger!: Logger;
+  private sockets: Socket[] = [];
+
+  public constructor(options: WebServer.Options) {
+    super(options);
+    if (options.port) this.port = options.port;
+    if (options.host) this.host = options.host;
+  }
+
+  /**
+   * Starts the http server after checking that the port is open
+   */
+  public async start(): Promise<void> {
+    try {
+      this.logger.debug('Starting web server');
+      await this.checkOsPort();
+      this.logger.debug(`Nothing listening on host: localhost port: ${this.port} - good!`);
+      this.server = http.createServer();
+
+      this.server.on('connection', (socket) => {
+        this.logger.debug(`socket connection initialized from ${socket.remoteAddress as string}`);
+        this.sockets.push(socket);
+      });
+      this.server.listen(this.port, this.host);
+    } catch (err) {
+      if (err.name === 'EADDRINUSE') {
+        const error = SfdxError.create('@salesforce/core', 'auth', 'PortInUse', ['PortInUseAction']);
+        error.actions = [messages.getMessage('PortInUseAction', [this.port])];
+        throw error;
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Closes the http server and all open sockets
+   */
+  public close(): void {
+    this.sockets.forEach((socket) => {
+      socket.end();
+      socket.destroy();
+    });
+
+    this.server.getConnections((_, num) => {
+      this.logger.debug(`number of connections open: ${num}`);
+    });
+
+    this.server.close();
+  }
+
+  /**
+   * sends a response error.
+   *
+   * @param statusCode he statusCode for the response.
+   * @param message the message for the http body.
+   * @param response the response to write the error to.
+   */
+  public sendError(status: number, message: string, response: http.ServerResponse): void {
+    response.statusMessage = message;
```

Comment:
We should eventually provide more information here. In the past, most people will see `localhost:1717` in the URL bar and an error like 404 and assume that the localhost is the problem because they don't have a server running. Either provide additional information or redirect them to a page that has more information on oauth failures.